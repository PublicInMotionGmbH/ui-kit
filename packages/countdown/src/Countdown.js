import React from 'react'
import PropTypes from 'prop-types'

import defaultFormat from './defaultFormat'

const propTypes = {
  /** Additional class name */
  className: PropTypes.string,

  /** Format of displayed date */
  render: PropTypes.func,

  /** The date to which it will count down. */
  targetDate: PropTypes.string.isRequired
}

const defaultProps = {
  render: defaultFormat
}

/**
 * Component which represents Countdown.
 *
 * @property {object} props
 * @property {string} [props.className]
 * @property {string} [props.format]
 * @property {string} props.targetDate
 * @class {React.Element}
 */
class Countdown extends React.PureComponent {
  state = {
    time: 0
  }

  /**
   * Set interval to update counter
   */
  componentDidMount () {
    this.tick()
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.targetDate !== this.props.targetDate) {
      this.tick()
    }
  }

  componentWillUnmount () {
    clearTimeout(this.tickTimeout)
  }

  tick = () => {
    const { targetDate } = this.props

    this.countTime(targetDate)

    clearTimeout(this.tickTimeout)
    this.tickTimeout = setTimeout(this.tick, 300)
  }

  /**
   * Set each part of date in state
   * @param {string} targetDate
   */
  countTime = (targetDate) => {
    const convertedDate = Date.parse(targetDate)

    this.setState({
      time: Math.max(0, convertedDate - Date.now())
    })
  }

  render () {
    const { time } = this.state
    const sec = Math.floor((time / 1000) % 60)
    const min = Math.floor((time / 1000 / 60) % 60)
    const hours = Math.floor(time / (1000 * 60 * 60) % 24)
    const days = Math.floor(time / (1000 * 60 * 60 * 24))
    const finished = !(sec + min + hours + days)
    const timeObj = {days, hours, min, sec, finished}

    return (
      <React.Fragment>
        {this.props.render(timeObj)}
      </React.Fragment>
    )
  }
}

Countdown.propTypes = propTypes
Countdown.defaultProps = defaultProps

export default Countdown
