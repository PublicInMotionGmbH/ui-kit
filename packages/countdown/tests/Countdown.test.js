import React from 'react'
import { shallow, mount } from 'enzyme'

import Countdown from '../src/Countdown'

// mock deadline in the future according to actual date
const setDeadline = (addTime) => {
  const dateNow = Date.now()
  const deadline = dateNow + addTime
  const deadlineConverted = new Date(deadline).toISOString()

  return deadlineConverted
}

describe('<Countdown />', () => {
  it('renders children correctly', () => {
    const wrapper = shallow(<Countdown targetDate={setDeadline(1000000000)} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('adds class when finished', () => {
    const wrapper = shallow(<Countdown targetDate={setDeadline(0)} />)
    expect(wrapper.find('.talixo-countdown').at(0).hasClass('talixo-countdown--finished')).toBe(true)
  })

  it('update component when new props are differents', () => {
    const wrapper = mount(<Countdown targetDate='2040-06-27T10:52:07.997Z' />)
    wrapper.setProps({
      targetDate: '2050-10-27T10:52:07.997Z'
    })
    expect(wrapper.prop('targetDate')).toBe('2050-10-27T10:52:07.997Z')
  })
})
