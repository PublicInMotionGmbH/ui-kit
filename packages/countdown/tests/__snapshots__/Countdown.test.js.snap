// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Countdown /> renders children correctly 1`] = `
<React.Fragment>
  <span
    className="talixo-countdown"
  >
    <span
      className="talixo-countdown talixo-countdown--days"
    >
      11
       
    </span>
    : 
    <span
      className="talixo-countdown talixo-countdown--hours"
    >
      13
       
    </span>
    : 
    <span
      className="talixo-countdown talixo-countdown--minutes"
    >
      46
       
    </span>
    : 
    <span
      className="talixo-countdown talixo-countdown--seconds"
    >
      39
       
    </span>
  </span>
</React.Fragment>
`;
